File: policy.py
"""Optional blocklist / allowlist: suffix-based matching on canonical DNS names."""

from __future__ import annotations

from pathlib import Path

import dns.name


def _parse_name_lines(path: Path) -> list[dns.name.Name]:
    names: list[dns.name.Name] = []
    for raw in path.read_text(encoding="utf-8").splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        n = dns.name.from_text(line)
        if not n.is_absolute():
            n = n.concatenate(dns.name.root)
        names.append(n.canonicalize())
    return names


def _is_under_or_equal(q: dns.name.Name, suffix: dns.name.Name) -> bool:
    if q == suffix:
        return True
    return q.is_subdomain(suffix)


class PolicyEngine:
    """
    - If allowlist is non-empty: only names matching at least one allow suffix may proceed.
    - Blocklist: names matching a block suffix return blocked (policy decision by caller).
    """

    def __init__(self, *, blocklist: list[dns.name.Name], allowlist: list[dns.name.Name]) -> None:
        self._block = sorted(blocklist, key=len, reverse=True)
        self._allow = sorted(allowlist, key=len, reverse=True)

    @classmethod
    def from_files(
        cls,
        blocklist_path: str | None,
        allowlist_path: str | None,
    ) -> PolicyEngine:
        blocks: list[dns.name.Name] = []
        allows: list[dns.name.Name] = []
        if blocklist_path:
            blocks = _parse_name_lines(Path(blocklist_path))
        if allowlist_path:
            allows = _parse_name_lines(Path(allowlist_path))
        return cls(blocklist=blocks, allowlist=allows)

    @classmethod
    def empty(cls) -> PolicyEngine:
        return cls(blocklist=[], allowlist=[])

    def is_blocked(self, qname: dns.name.Name) -> bool:
        n = qname
        if not n.is_absolute():
            n = n.concatenate(dns.name.root)
        n = n.canonicalize()

        if self._allow:
            if not any(_is_under_or_equal(n, a) for a in self._allow):
                return True

        return any(_is_under_or_equal(n, b) for b in self._block)

File: web/lib/policy.ts
import { isUnderOrEqual, normalizeFqdn } from "./dns-names";

function parseNameLines(lines: string[]): string[] {
  const out: string[] = [];
  for (const raw of lines) {
    const line = raw.trim();
    if (!line || line.startsWith("#")) continue;
    out.push(normalizeFqdn(line));
  }
  return out;
}

export class PolicyEngine {
  private readonly _block: string[];
  private readonly _allow: string[];

  constructor(blocklist: string[], allowlist: string[]) {
    this._block = [...blocklist].sort((a, b) => labelsLen(b) - labelsLen(a));
    this._allow = [...allowlist].sort((a, b) => labelsLen(b) - labelsLen(a));
  }

  static empty(): PolicyEngine {
    return new PolicyEngine([], []);
  }

  static fromLines(blockLines: string[], allowLines: string[]): PolicyEngine {
    return new PolicyEngine(parseNameLines(blockLines), parseNameLines(allowLines));
  }

  isBlocked(qname: string): boolean {
    const n = normalizeFqdn(qname);

    if (this._allow.length > 0) {
      if (!this._allow.some((a) => isUnderOrEqual(n, a))) return true;
    }

    return this._block.some((b) => isUnderOrEqual(n, b));
  }
}

function labelsLen(fqdn: string): number {
  return fqdn.replace(/\.$/, "").split(".").filter(Boolean).length;
}
